### ISSUE TITLE
Unable to connect to Logstash server by Logstash Forwarder due to problem with TLS/SSL

### ISSUE THREAD

**Issue by paoolo** [elastic/logstash#2328]
Logstash input configuration:

    input {
      lumberjack {
        port => 6000
        type => "logs"
        ssl_certificate => "/etc/pki/tls/certs/logstash-forwarder.crt"
        ssl_key => "/etc/pki/tls/private/logstash-forwarder.key"
      }
    }

Logstash Forwarder configuration:

    "network": {
        "servers": ["IP:6000"],
        "timeout": 15,
        "ssl ca": "/etc/pki/tls/certs/logstash-forwarder.crt"
    },

Error:

    Failed to tls handshake with IP tls: server selected unsupported protocol version 301

**Comment by dariuszmysliwiec** [comment]
I experience exactly the same problem. It's strange because around 1 month ago I compiled few logstash-forwarders instances and they were working.

**Comment by jordansissel** [comment]
What jvm and version of java are you using? What version of logstash?

I'm confused. I tested https://github.com/elasticsearch/logstash-forwarder/commit/22d672da1fd6f670945fd8754632ebd4d8212ae8 before committing  and it works.

**Comment by dariuszmysliwiec** [comment]
My logstash was 1.4.2-modified. Updating my right away!

**Comment by jordansissel** [comment]
yeah I may have to fix lsf master to not have tls 1.1

**Comment by driskell** [comment]
301 is TLS 1.0

You need ony disable SSLv3 for poodle don't you? Is there a reason to disallow TLS 1.0?
Might be the jvm/is combination doesn't support 1.1 as you point out.

**Comment by jordansissel** [comment]
@driskell in my testing, jruby/java/go don't seem to negotiate properly. I'm not entirely sure why.

lsf master -> Logstash 1.4.2 fails to connect due to lsf's minversion constraint. It's quite weird. I still need to debug why.

**Comment by jordansissel** [comment]
I've reverted the TLS 1.1 change until I can figure out what's going on.

**Comment by jordansissel** [comment]
I can confirm that:

*    Logstash master server hello reports TLS v1.1
 *   Logstash 1.4.2 server hello reports TLS v1.0

**Comment by maxhyjal** [comment]
Just out of curiosity, why is TLS v1.2 not supported(or going to be supported) on the logstash server side?

**Comment by jordansissel** [comment]
@maxhyjal TLS v1.2 is not yet supported by JRuby (or Java, I am not sure yet).

**Comment by jordansissel** [comment]
Java supports TLS v1.2 it looks like, could be something in bounceycastle or jruby-openssl limiting it to v1.1

**Comment by rogerthat** [comment]
as far as i can see it's not an java-issue (tested 1.7_25 up to 1.8_45, jdk as well as jre, different switches like 
-Dhttps.protocols='TLSv1,TLSv1.1,TLSv1.2' -Ddeployment.security.SSLv3=false  -Ddeployment.security.TLSv1=true  -Ddeployment.security.TLSv1.1=true -Ddeployment.security.TLSv1.2=true 

w/out beeing able to switch on tlsv1.1/1.2; seems more like a jruby - problem here.

**Comment by jordansissel** [comment]
Yep. I wrote some test code to verify TLS v1.2 in via javax.net.ssl. Problem is either jruby-openssl or bounceycastle. I'm looking into it.

**Comment by jordansissel** [comment]
Fixing TLS v1.2 support will require a few improvements to jruby-openssl. I'll do this soon, but not now.

**Comment by jordansissel** [comment]
Current task is to figure out how to support both TLS 1.1 and TLS 1.0

**Comment by maxhyjal** [comment]
Will Logstash 1.5 include support for TLS 1.1 and TLS 1.2?

**Comment by driskell** [comment]
I've been investigating jruby-openssl and the default protocol list is the same as MRI OpenSSL gem, "SSLv23". With MRI this includes TLS 1.1 and 1.2 if your OpenSSL library supports it.

jruby-openssl is not OpenSSL but uses JSSE so it needs to translate everything. As such it only currently translates SSLv23 to include TLS 1.0.

@jordansissel Do you know why Logstash master would report TLS 1.1? I can't see how it is possible unless the SSLContext is modified so the protocol version is TLSv1.1 or TLS, both being jruby-openssl specific (the OpenSSL equivalent is TLSv1_1 only.)

I have a branch that enables TLSv1.1 and TLSv1.2 if available when the protocol is set to SSLv23 which is what happens in MRI with latest OpenSSL. Problem is none of the new TLS 1.2 ciphers are supported and adding those is not trivial. Just waiting on my current SSLSocket ssl_version method to be merged before I start discussing with them.

TLDR: Full TLS 1.2 in jruby is some time off. Protocol only support (with none of the new 1.2 ciphers) might be possible soon though.

**Comment by driskell** [comment]
Also to disable SSLv3 etc there's SSL::OP_NO_SSLv2 and SSL::OP_NO_SSLv3 which work a charm and seem to be the general way of disabling old protocols whilst allowing newer ones to be added without application modification.

**Comment by jordansissel** [comment]
I think the original problem was resolved here (TLS v1.1 requirement in one commit of logstash-forwarder that was rolled back to TLS 1.0)

**Comment by suyograo** [comment]
LS 1.5.3 explicitly removed support for SSLv3. Closing this
